src/audio/pipewire.c

```diff
@@ -0,0 +1,322 @@
+    char *sink_name;             // the name of the sink, as given by speech dispatcher
+    int32_t eventfd_number;      // used in on_process to signal the thread where pipewire_play is running that the ringbuffer has been drained to the point where playback either finished or is in progress, but for sure to the point where we could push more audio, because our side of the buffer at least is drained
+} module_state;
+
+// pipewire on process callback
+// this function would be called each time the server requests data from us
+static void on_process(void *userdata)
+{
+    module_state *state = (module_state *)userdata;
+    // a lot of the code below has been taken from the playing a tone pipewire tutorial, see the pipewire documentation for more details
+
+    struct pw_buffer *b;
+    struct spa_buffer *buf;
+    uint32_t *destination_memory, read_index, available_for_loading, must_load_with_silence;
+    int32_t available_samples, number_of_samples;
+    if ((b = pw_stream_dequeue_buffer(state->stream)) == NULL)
+    {
+        pw_log_warn("out of buffers: %m");
+        return;
+    }
+    buf = b->buffer;
+    // pipewire's buffers are mapped directly from the server to the client
+    // As such, if the memory we're given doesn't exist, we have to abort imediately, because doing otherwise most certainly would lead to undefined behaviour
+    if ((destination_memory = buf->datas[0].data) == NULL)
+        return;
+    // in the process callback, we take the samples from the ring buffer we used in the pipewire_play function and send it to pipewire for playback
+    // this will be done using the following two step algorythm:
+    // we first get the read index of the ringbuffer, which also returns the amount available in there
+    // if we have enough data in the ringbuffer, aka if the read index isn't less than requested, then we load all of it at once in pipewire
+    // however, if the remainder between the filled part of the buffer and the value of requested is greatter than 0, then we fill that part of the buffer with silence, while taking everything else, hoping we will have more samples next time
+    available_samples = spa_ringbuffer_get_read_index(&state->rb, &read_index);
+    number_of_samples = buf->datas[0].maxsize; // how many samples, in bytes,  of audio are in this chunk. Technically, requested itself could be used, but this is done instead to avoid some very rare, but important, edge cases, for example buggy hardware drivers which don't probe the card correctly, so on_process gets called very often, with a value of requested samples greatter than what pipewire could allocate, in maxsize
+#if PW_CHECK_VERSION(0, 3, 49)
+    if (b->requested > 0)
+    {
+        number_of_samples = SPA_MIN(number_of_samples, b->requested * state->stride); // because this value is given in frames, and we want samples
```

Comment:
`samples` is a synonym of `frame`. You need to use `bytes` instead of `samples`.

Reply:
I meant fix the comment